host/sgx/windows/switchless.c

```diff
@@ -13,18 +13,37 @@ void oe_host_worker_wait(oe_host_worker_context_t* context)
     int32_t oldval = 1;
     int32_t newval = 0;
 
-    if (_InterlockedCompareExchange((long*)&context->event, newval, oldval) ==
-        0)
+    if (_InterlockedCompareExchange(event, newval, oldval) == 0)
```

Comment:
Going through the call sequence, the "event" passed in to this function is allocated via calloc() - 
host_contexts = calloc(...) and enclave_contexts = calloc(...). Is that correct?

The reason I ask is because the documentation for the _Interlocked* APIs says the memory needs to be aligned (32bit aligned on x86 and 64bit aligned on x64).

Reply:
@fchinchilla

Thanks for reviewing.

Yes, we allocate host_contexts and enclave_contexts via calloc.

Due to malloc's [guarantee](https://en.cppreference.com/w/c/memory/malloc) to return an address suitably aligned for any fundamental type, implying at least 64 bit alignment (alignment of void*) if not 128-bit, and due to `int32_t event` [guaranteed](https://github.com/anakrish/openenclave/blob/f8973684d0799cdc34c39ba29cf3163213ecad7b/include/openenclave/internal/switchless.h#L35) to be at a 32-bit aligned-offset (20 bytes) within its calloc-ed enclosing structure, we can be sure that `event` is always 32-bit aligned.

Does x64 require that 32-bit values are also aligned to 64 bits for interlocked operations to work?

Reply:
Looks like this is OK then. The _InterlockedCompareExchange64 needs 64bit aligned, but _InterlockedCompareExchange only needs 32bit aligned. The other usages of Interlocked APIs appear to be the generic and 32bit alignment should be ok.

Reply:
Thanks for confirming. I was worried for bit.